### DISCUSSION TITLE
Modify the "microsoft-teams-emergency-operations-center"

### DISCUSSION THREAD

**Discussion by mvcsharepointdev** [community/community#141722] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

Question

### Body

I am working on a new project and the customer have already installed and used this Teams tab from GitHub @ https://github.com/OfficeDev/microsoft-teams-emergency-operations-center enter image description here

Now they want to do some modifications to this solution, such as :-

1) Change some titles for the fields
2) Currently when an item get created it provision some channels on teams, so they want to modify the name of those channels and provision more channels.

So my first question is about how i need to do those changes,, are those steps valid:-

1) Clone the code from the above github link
2) Modify the code accordingly
3) Deploy and publish the solution to get a new .sppkg solution and re-upload it to the sharepoint app

First Question, are those valid steps to do the modifications? i built a lot of SPFx solutions, but never modify existing one on GitHub.??

Second Question, when the customer install this solution for the first time, a new site and lists were provisioned... so if i replace the old .sppkg file with the new one, will this re-provision the site and the lists? or this will only happen when we installed this solution for the first time only?

Third question, now if i do those modifications inside our local version, and Microsoft release a new version of this app inside GitHub, and we want to have this new version, then our local modification done previously will get lost? am i correct? so can we for example post those changes to GitHub as well, so they will be part of the new releases done by Microsoft ?

Thanks

**Comment by MorelliniMatteo** [comment]
1. Valid Steps for Modifications
Your outlined steps for modifying the GitHub solution are generally correct. Here’s a refined process:

Clone the Code:

Use Git to clone the repository from the provided link.
Modify the Code:

Change the titles of the fields as needed.
Modify the logic for channel creation (changing names and adding more channels) in the codebase.
Make sure to update any configurations related to these changes in the appropriate files.
Deploy and Publish the Solution:

Run the appropriate commands to build the solution (e.g., gulp build, gulp bundle, gulp package-solution).
This will generate a new .sppkg file in the sharepoint/solution folder.
Upload the updated .sppkg file to your SharePoint App Catalog and ensure it is deployed.
Conclusion: Yes, these steps are valid for modifying the existing solution.

2. Re-provisioning Sites and Lists
When you replace the existing .sppkg file with a new one, the behavior concerning provisioning of sites and lists will depend on how the original solution was designed. Generally:

Initial Installation: When the solution is installed for the first time, it typically provisions sites and lists based on the code and configuration provided in the solution.

Subsequent Updates: If you simply replace the .sppkg file:

If the code is designed to check for existing sites/lists: It may skip re-provisioning if it detects that they already exist.
If the code does not handle existing provisions: It might attempt to create them again, which can lead to errors.
Conclusion: Generally, replacing the .sppkg file will not automatically re-provision the sites and lists unless explicitly coded to do so.

3. Handling Future Updates and Local Modifications
If you make modifications to the local version of the app and then a new version is released by Microsoft, here are some considerations:

Local Changes vs. Upstream Updates: If you update to the new version directly from GitHub, your local changes will likely be lost unless you have a systematic way of merging your changes back into the updated codebase.

Version Control: To manage local modifications:

Consider creating a separate branch in your local Git repository for your changes. This way, you can later merge in updates from the main repository without losing your modifications.
Contributing Back to GitHub:

If your changes are useful to others, consider forking the repository, making your changes, and then submitting a pull request (PR) to the original repository. This way, your modifications could potentially be included in future releases, provided they meet the project’s contribution guidelines.
Conclusion: Yes, without a proper merge strategy, local modifications can be lost. Maintaining a separate branch for your changes and potentially contributing back to the original repository is a solid approach.

**Reply by mvcsharepointdev** [reply]
@MorelliniMatteo Thanks a lot for your very useful reply, i do not think we will publish our modifications to the UpStream as those modifications are very customized to our requirements,  even if we do so, i am sure not all of them will get accepted.  so i read about something called Fork the repository. where i will Fork the upstream repo >> do my modifications >> then when the upstream do a new release >> i will merge the changes >> and `git merge` will mark them and you'll have to resolve them manually.. what do you think? are those the way to go?

**Reply by MorelliniMatteo** [reply, marked as answer]
Your approach to forking the repository for your modifications is sound and follows best practices in version control. Here’s a breakdown of your proposed workflow and some additional insights:

Workflow for Forking and Modifying the Repository
Fork the Repository:

Start by forking the upstream repository on GitHub. This creates a copy of the repository under your own GitHub account, allowing you to make changes without affecting the original project.
Clone Your Fork:

Clone your forked repository to your local machine. This allows you to work on the modifications locally.
Make Your Modifications:

Implement the custom changes required for your project. This could involve changing titles, modifying channel creation logic, and updating any relevant configurations.
Commit and Push Changes:

Once your modifications are complete, commit your changes and push them to your forked repository on GitHub.
Stay Updated with Upstream Changes:

To keep your fork up to date with the upstream repository:
Add the Upstream Remote: Add the original repository as an upstream remote using the command:
bash command -- git remote add upstream

Fetch Upstream Changes: Regularly fetch changes from the upstream repository:
bash command -- git fetch upstream

Merge Upstream Changes: When there’s a new release in the upstream repository, merge those changes into your fork:
bash command -- git checkout main # Switch to your main branch
git merge upstream/main # Merge upstream changes into your local branch

Resolve Merge Conflicts:

If there are changes in the upstream repository that conflict with your modifications, Git will mark them as conflicts. You'll need to manually resolve these conflicts by editing the affected files, then staging and committing the resolved files:
bash command -- git add ...
git commit -m "Resolved merge conflicts"

**Reply by MorelliniMatteo** [reply]
please, if you can, mark my answer as answered.

**Reply by mvcsharepointdev** [reply]
@MorelliniMatteo  Thanks a lot

**Reply by MorelliniMatteo** [reply]
thanks to you, have a nice day
